### ISSUE TITLE
many0! etc. return IResult::Done with empty input

### ISSUE THREAD

**Issue by dlesl** [Geal/nom#650]
### Problem
I'm writing a streaming parser which needs to behave the same way regardless of how the incoming data arrives. If I have a slice for which `parser(slice)` returns `IResult::Done([], _)`, then `parser(&slice[..n])` needs to return `IResult::Incomplete(_)` for every value of n < slice.len(). This seems to be the way all the nom parsers behave with the exception of `many0!`, `fold_many0!` etc.
### Example
```rust
named!(hello<&[u8]>, tag!("hello"));
assert_eq!(hello(b""), IResult::Incomplete(Needed::Size(5)));
// OK

named!(hello_many0<Vec<&[u8]>>, many0!(hello));
assert_eq!(hello_many0(b""), IResult::Incomplete(Needed::Size(5)));
// Fails, parser returns `Done([], [])`
```
### Expected behaviour
This is a problem because if the data happens to arrive in such a way that input is empty just before the `many0!` parser, the final result is different than if one more byte had been available. Instead, `many0!` should pass any input (even empty) on to the child parser, and only return `Done` when the child parser returns `Error`

By commenting out the `if input.input_len() == 0` check in `many0!` I was able to get my parser to work the way I wanted, but I'm not sure if this wouldn't break other parsers. I suspect in some cases people would need to wrap the child parser in `complete!`. Or maybe I'm using `many0!` wrong? ;)

**Comment by Geal** [comment]
this looks like an issue, yes, I'll think about it

**Comment by Geal** [comment]
Hello, this should be fixed in 6d5b236

**Comment by dlesl** [comment]
Thanks, it works! (for my parser)
